rllib/agents/alpha_star/alpha_star.py

```diff
@@ -0,0 +1,456 @@
+        if self.config["max_num_policies_to_train"] is None:
+            self.config["max_num_policies_to_train"] = len(
+                self.workers.local_worker().get_policies_to_train()
+            )
+
+        # Single CPU replay shard (co-located with GPUs so we can place the
+        # policies on the same machine(s)).
+        ReplayActor = ray.remote(
+            num_cpus=1,
+            num_gpus=0.01
+            if (self.config["num_gpus"] and not self.config["_fake_gpus"])
+            else 0,
+        )(MixInMultiAgentReplayBuffer)
+
+        # Setup remote replay buffer shards and policy learner actors
+        # (located on any GPU machine in the cluster):
+        replay_actor_args = [
+            self.config["replay_buffer_capacity"],
+            self.config["replay_buffer_replay_ratio"],
+        ]
+
+        # Create a DistributedLearners utility object and set it up with
+        # the initial first n learnable policies (found in the config).
+        distributed_learners = DistributedLearners(
+            config=self.config,
+            max_num_policies_to_train=self.config["max_num_policies_to_train"],
+            replay_actor_class=ReplayActor,
+            replay_actor_args=replay_actor_args,
+        )
+        for pid, policy_spec in ma_cfg["policies"].items():
+            if pid in self.workers.local_worker().get_policies_to_train():
+                distributed_learners.add_policy(pid, policy_spec)
+
+        # Store distributed_learners on all RolloutWorkers
+        # so they know, to which replay shard to send samples to.
+
+        def _set_policy_learners(worker):
+            worker._distributed_learners = distributed_learners
+
+        ray.get(
+            [
+                w.apply.remote(_set_policy_learners)
+                for w in self.workers.remote_workers()
+            ]
+        )
+
+        self.distributed_learners = distributed_learners
+
+        # Store the win rates for league overview printouts.
+        self.win_rates: DefaultDict[PolicyID, float] = defaultdict(float)
+
+    @override(Trainer)
+    def step(self) -> ResultDict:
+        # Perform a full step (including evaluation).
+        result = super().step()
+
+        # Based on the (train + evaluate) results, perform a step of
+        # league building.
+        self.league_builder.build_league(result=result)
+
+        return result
+
+    @override(Trainer)
+    def training_iteration(self) -> ResultDict:
+        # Trigger asynchronous rollouts on all RolloutWorkers.
+        # - Rollout results are sent directly to correct replay buffer
+        #   shards, instead of here (to the driver).
+        with self._timers[SAMPLE_TIMER]:
+            sample_results = asynchronous_parallel_requests(
+                remote_requests_in_flight=self.remote_requests_in_flight,
+                actors=self.workers.remote_workers() or [self.workers.local_worker()],
+                ray_wait_timeout_s=0.01,
```

Comment:
There feels like there is some downside to manually setting the timeouts for sampling and training -- isn't this going to be dependent on rollout length and training time?

Reply:
I had a gotcha moment about this too.
I think this timeout is basically saying: give me whatever is ready right now, so I can continue processing them. leave the un-finished remote calls for the next round of iteration.

Reply:
"is basically saying: give me whatever is ready right now":
Yes, correct.

I think we should try different things here, but probably timeout=0.0 is the best value here.